File: src/backend/youtubeApi.ts
import {getYoutube} from "./configureYoutube";

// https://developers.google.com/youtube/v3/docs/videos#liveStreamingDetails.activeLiveChatId
export async function getLiveChatId(videoId: string) {
    const response = await getYoutube().videos.list({
        part: ['liveStreamingDetails'],
        id: [videoId],
    });

    const video = response.data.items?.[0];
    return video?.liveStreamingDetails?.activeLiveChatId;
}

// https://developers.google.com/youtube/v3/live/docs/liveChatMessages
export async function getLiveChatMessages(liveChatId: string, pageToken?: string) {
    const response = await getYoutube().liveChatMessages.list({
        liveChatId,
        part: ['snippet', 'authorDetails'],
        pageToken,
    });

    return {
        messages: response.data.items || [],
        nextPageToken: response.data.nextPageToken,
        pollingIntervalMillis: response.data.pollingIntervalMillis,
    };
}

// https://developers.google.com/youtube/v3/live/docs/liveChatMessages
export async function getLiveChatMessagesNoPageToken(liveChatId: string) {
    const response = await getYoutube().liveChatMessages.list({
        liveChatId,
        part: ['snippet', 'authorDetails'],
    });

    return response.data.items?.map(item => ({
        author: item.authorDetails?.displayName || 'Unknown',
        message: item.snippet?.displayMessage || '',
        eventType: item.snippet?.type || '',
        textDetails: item.snippet?.textMessageDetails?.messageText || ''
    })) || [];
}

File: backend/myyoutube/youtubeAPI.go
package myyoutube

import (
	"fmt"

	"google.golang.org/api/youtube/v3"
)

func GetLiveChatID(service *youtube.Service, videoID string) (string, error) {
	// Get the live broadcast details
	broadcastCall := service.Videos.List([]string{"liveStreamingDetails"}).Id(videoID)
	broadcastResponse, err := broadcastCall.Do()
	if err != nil {
		return "", fmt.Errorf("error retrieving live broadcast details: %v", err)
	}

	// Check if live chat ID is available
	if len(broadcastResponse.Items) > 0 {
		liveChatID := broadcastResponse.Items[0].LiveStreamingDetails.ActiveLiveChatId
		if liveChatID != "" {
			return liveChatID, nil
		}
	}

	return "", fmt.Errorf("no live broadcast found or no active live chat")
}

// Response structure for live chat messages
type LiveChatMessagesResponse struct {
	Messages              []*youtube.LiveChatMessage
	NextPageToken         string
	PollingIntervalMillis int64
}

func GetLiveChatMessagesAsync(
	service *youtube.Service,
	liveChatID string,
	pageToken string) (chan []*youtube.LiveChatMessage, chan string, chan int64, chan error) {

	messagesChan := make(chan []*youtube.LiveChatMessage)
	nextPageTokenChan := make(chan string)
	pollingIntervalMillisChan := make(chan int64)
	errorChan := make(chan error)

	go func() {
		call := service.LiveChatMessages.List(liveChatID, []string{"snippet", "authorDetails"})
		if pageToken != "" {
			call = call.PageToken(pageToken)
		}

		response, err := call.Do()
		if err != nil {
			errorChan <- fmt.Errorf("error retrieving live chat messages: %v", err)
			return
		}

		messagesChan <- response.Items
		nextPageTokenChan <- response.NextPageToken
		pollingIntervalMillisChan <- response.PollingIntervalMillis
	}()

	return messagesChan, nextPageTokenChan, pollingIntervalMillisChan, errorChan
}

func GetLiveChatMessages(
	service *youtube.Service,
	liveChatID string,
	pageToken string) ([]*youtube.LiveChatMessage, string, int64, error) {

	call := service.LiveChatMessages.List(liveChatID, []string{"snippet", "authorDetails"})
	if pageToken != "" {
		call = call.PageToken(pageToken)
	}

	response, err := call.Do()
	if err != nil {
		return nil, "", 0, fmt.Errorf("error retrieving live chat messages: %v", err)
	}

	return response.Items, response.NextPageToken, response.PollingIntervalMillis, nil
}
